### PR TITLE
KNIP-678: Automated recovery from failed nodes - Reactive -IPI

### PR DESCRIPTION
**Pull request by PrasadDesala**
This patch automates KNIP-678: Automated recovery from failed nodes. Rective-AWS-IPI case.
It covers two failures,
1) Power off and then terminate the instance
2) Terminating a running instance.

Signed-off-by: Prasad Desala <tdesala@redhat.com>

### CODE REVIEW

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        )
+        log.info(
+            f"{osd_node_obj.name} associated machineset is {machineset_name}"
+        )
+
+        # Add a new node and label it
+        add_new_node_and_label_it(machineset_name)
+
+        # Induce failure
+        aws_obj = platform_nodes.AWSNodes()
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: should be used using the 'nodes' fixture from tests/conftest.py

see reference here https://github.com/red-hat-storage/ocs-ci/blob/1839221c2847b5f0d8dbe287e6f0a5f0542c8bec/tests/manage/z_cluster/nodes/test_nodes_restart.py#L71

Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Done

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                resource=pod_obj, state=constants.STATUS_RUNNING, timeout=300
+            )
+
+        # Check basic cluster functionality by creating resources
+        # (pools, storageclasses, PVCs, pods - both CephFS and RBD),
+        # run IO and delete the resources
+        self.sanity_helpers.create_resources(pvc_factory, pod_factory)
+
+        # Perform cluster and Ceph health checks
+        self.sanity_helpers.health_check()
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Please check https://github.com/red-hat-storage/ocs-ci/pull/1345/files#r368893770 and implement the required parts here.

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            machine_name
+        )
+        log.info(
+            f"{osd_node_obj.name} associated machineset is {machineset_name}"
+        )
+
+        # Add a new node and label it
+        add_new_node_and_label_it(machineset_name)
+
+        # Induce failure
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Before inducing the failure, are running dc pods and IOs from it? Please make sure you have few dc pods running IOs along with the supported workloads like pgsql, metrics, logging and registry....these IOs should not fail 'unexpectedly'

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        if failure == "shutdown":
+            nodes.stop_nodes([osd_node_obj], wait=True)
+            log.info(f"Successfully powered off node: {osd_node_obj.name}")
```
Comment (prsurve) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: This scenario is normal shutdown, not abrupt shutdown

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
```
Comment (prsurve) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: to introduce abrupt failure you can stop network interface

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py (hunk 2 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Done

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py (hunk 3 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Done

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py (hunk 4 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Done

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py (hunk 5 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Done

File: tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                        log.info(
+                            "BZ 1789419 - WA executed since canary pod stuck"
+                            " at pending state"
+                        )
+
+        # Check basic cluster functionality by creating resources
+        # (pools, storageclasses, PVCs, pods - both CephFS and RBD),
+        # run IO and delete the resources
+        self.sanity_helpers.create_resources(pvc_factory, pod_factory)
+
```
Comment (prsurve) [inline tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: delete_resource function is missing

Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_automated_recovery_from_failed_nodes_reactive_IPI.py]: Added

File: ocs_ci/ocs/resources/pod.py [hunk 7]
```diff
@@ -1176,3 +1176,20 @@ def get_noobaa_pods(noobaa_label=constants.NOOBAA_APP_LABEL, namespace=None):
...
+
+def wait_for_dc_app_pods_to_reach_running_state(dc_pod_obj):
+    """
+    Wait for DC app pods to reach running state
+
+    Args:
+        dc_pod_obj (list): list of dc app pod objects
+    """
+    for pod_obj in dc_pod_obj:
+        name = pod_obj.name[:-8]
```
Comment (prsurve) [inline ocs_ci/ocs/resources/pod.py]: you can get pod name from pod_obj.get_labels().get('name')

Comment (PrasadDesala) [reply ocs_ci/ocs/resources/pod.py]: Done

**Comment by PrasadDesala** [comment]
Closing this old PR. Will open a new one

**Pull request closed without merging**